src/apps/shared/ordering.py

```diff
@@ -26,40 +63,136 @@ class Clause:
+            if parsed_field is None:
+                continue
+            _direction, field = parsed_field
+            if not self._is_sql_field(field):
+                continue
+            clause = self._prepare_sql_clause(parsed_field)
+            if clause is not None:
+                clauses.append(clause)
+        return clauses
+
+    # Returns list of fields that this Ordering class supports for sorting, based on whether
+    # to use manual sorting or not. Returns field names in camelCase suitable for API output.
+    def get_ordering_fields(self, use_manual_sorting: bool = False):
+        use_manual_sorting = use_manual_sorting and bool(self.manual_fields)
+        fields = (self.manual_fields if use_manual_sorting else self.sql_fields).keys()
+        return list(to_camelcase(word) for word in fields)
+
+    def _is_sql_field(self, field: str):
+        return field in self.sql_fields
 
-    def _prepare_ordered_field(self, value: str):
+    def _is_manual_field(self, field: str):
+        return field in self.manual_fields
+
+    def _parse_ordered_field(self, value: str):
         if not value:
             return None
-        sign, field = value[0], value[1:]
-        if sign != "-":
+        direction, field = value[0], value[1:]
+        if direction != "-":
             field = value
-            sign = "+"
+            direction = "+"
         field = field.lstrip("+")
+        return (direction, field)
 
-        if field not in self.fields:
+    def _prepare_sql_clause(self, value: tuple[OrderingDirection, str]):
+        direction, field = value
+        if not self._is_sql_field(field):
             return None
 
-        return self.actions[sign](self.fields[field])
+        return self.actions[direction](self.sql_fields[field])
+
+    def manual_sort(self, data: list, fields: list[tuple[OrderingDirection, OrderingField]]):
+        def comparer(left, right):
+            for direction, field in fields:
+                left_value = right_value = None
+                # Extract array field value
+                if isinstance(field, tuple):
+                    field, index = field
+                    left_value = left[field][index]
+                    right_value = right[field][index]
+                # Extract scalar field value
+                else:
+                    left_value = left[field]
+                    right_value = right[field]
+
+                # Protect against None values (None ranks lowest in SQL by default)
+                if left_value is None and right_value is None:
+                    continue
+                elif left_value is None:
+                    # Set left value to maximum possible value for the type
+                    if isinstance(right_value, (int, float)):
+                        left_value = float("inf")  # maximum number
+                    elif isinstance(right_value, str):
+                        left_value = chr(1114111)  # maximum string
```

Comment:
Please do not use this value as magic number

Reply:
Sure, I'll see if I can come up with different logic that doesn't involve magic numbers.